rust/main/agents/relayer/src/server/message_retry.rs

```diff
@@ -1,32 +1,89 @@
+use std::sync::Arc;
+
 use crate::settings::matching_list::MatchingList;
+
 use axum::{extract::State, routing, Json, Router};
+
 use derive_new::new;
-use tokio::sync::broadcast::Sender;
+use serde::{Deserialize, Serialize};
+use tokio::sync::{broadcast::Sender, mpsc, Mutex};
 
 const MESSAGE_RETRY_API_BASE: &str = "/message_retry";
 
-#[derive(new, Clone)]
+#[derive(new)]
 pub struct MessageRetryApi {
-    tx: Sender<MatchingList>,
+    tx: Sender<MessageRetryRequest>,
+    rx: Arc<Mutex<mpsc::Receiver<MessageRetryResponse>>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct MessageRetryApiState {
+    pub tx: Sender<MessageRetryRequest>,
+    pub rx: Arc<Mutex<mpsc::Receiver<MessageRetryResponse>>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct MessageRetryRequest {
+    pub uuid: String,
+    pub pattern: MatchingList,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct MessageRetryResponse {
+    /// ID of the retry request
+    pub uuid: String,
+    /// how many pending operations were processed
+    pub processed: usize,
+    /// how many of the pending operations matched the retry request pattern
+    pub matched: u64,
 }
 
 async fn retry_message(
-    State(tx): State<Sender<MatchingList>>,
+    State(state): State<MessageRetryApiState>,
     Json(retry_req_payload): Json<MatchingList>,
-) -> String {
-    match tx.send(retry_req_payload) {
-        Ok(_) => "Moved message(s) to the front of the queue".to_string(),
-        // Technically it's bad practice to print the error message to the user, but
-        // this endpoint is for debugging purposes only.
-        Err(err) => format!("Failed to send retry request to the queue: {}", err),
+) -> Result<Json<MessageRetryResponse>, String> {
+    let uuid = uuid::Uuid::new_v4();
+    let uuid_string = uuid.to_string();
+
+    tracing::debug!("Sending message retry request: {uuid_string}");
+
+    state
+        .tx
+        .send(MessageRetryRequest {
+            uuid: uuid_string.clone(),
+            pattern: retry_req_payload,
+        })
+        .map_err(|err| {
+            // Technically it's bad practice to print the error message to the user, but
+            // this endpoint is for debugging purposes only.
+            format!("Failed to send retry request to the queue: {}", err)
+        })?;
+
+    let mut rx = state.rx.lock().await;
+
+    // Wait for response from message retry.
+    // Warning: this potentially blocks other retry requests
```

Comment:
I wonder if we should use the lock that we acquire above as a way to prevent this? i.e. by acquiring the lock before we do the `tx.send`.

It's a bit of a weird and clunky hammer, but imo the usage of this is infrequent and only by the relayer operator, so I'm not really concerned about the poor UX / practice not allowing concurrent retry requests. I don't have a strong intuition for how quick we expect to get an ack back here, I assume we can end up blocking for many seconds if prep queues are large

Reply:
Actually a maybe better idea if we wanna avoid holding locks for this -- couldn't we have `MessageRetryRequest` include a channel that's used for an ack back to this specific request?

Reply:
You mean like a oneshot channel inside `MessageRetryRequest`?

Reply:
yeah exactly

oneshot may or may not be the right channel bc of the whole multiple submitters needing to send a response back thing

But I think this is probably a better design? wdyt

Reply:
Yeah, sounds good to me. Just generate a mpsc channel for every retry request. And that channel will be solely used for that specific retry request.

Reply:
not sure if memory usage assumptions were considered here but I think we should be ok given the very low volume of retry requests